core/src/main/java/kafka/server/share/SharePartitionManager.java

```diff
@@ -328,6 +366,202 @@ public void close() throws Exception {
         // TODO: Provide Implementation
     }
 
+    private ShareSessionKey shareSessionKey(String groupId, Uuid memberId) {
+        return new ShareSessionKey(groupId, memberId);
+    }
+
+    private static String partitionsToLogString(Collection<TopicIdPartition> partitions) {
+        return FetchSession.partitionsToLogString(partitions, log.isTraceEnabled());
+    }
+
+    /**
+     * Recursive function to process all the fetch requests present inside the fetch queue
+     */
+    private void maybeProcessFetchQueue() {
+        if (!processFetchQueueLock.compareAndSet(false, true)) {
+            // The queue is already being processed hence avoid re-triggering.
+            return;
+        }
+
+        // Initialize the topic partitions for which the fetch should be attempted.
+        Map<TopicIdPartition, FetchRequest.PartitionData> topicPartitionData = new LinkedHashMap<>();
+        ShareFetchPartitionData shareFetchPartitionData = fetchQueue.poll();
+        if (shareFetchPartitionData == null) {
+            // No more requests to process, so release the lock. Though we should not reach here as the lock
+            // is acquired only when there are requests in the queue. But still, it's safe to release the lock.
+            releaseProcessFetchQueueLock();
+            return;
+        }
+
+        try {
+            shareFetchPartitionData.topicIdPartitions.forEach(topicIdPartition -> {
+                SharePartitionKey sharePartitionKey = sharePartitionKey(
+                    shareFetchPartitionData.groupId,
+                    topicIdPartition
+                );
+                SharePartition sharePartition = partitionCacheMap.computeIfAbsent(sharePartitionKey,
+                    k -> new SharePartition(shareFetchPartitionData.groupId, topicIdPartition, maxInFlightMessages, maxDeliveryCount,
+                        recordLockDurationMs, timer, time, persister));
+                int partitionMaxBytes = shareFetchPartitionData.partitionMaxBytes.getOrDefault(topicIdPartition, 0);
+                // Add the share partition to the list of partitions to be fetched only if we can
+                // acquire the fetch lock on it.
+                if (sharePartition.maybeAcquireFetchLock()) {
```

Comment:
> However if have any partition acquired then the request will be completed with response of locked partition records.

That's true. However, if there is no new data coming into the locked partitions, but there is data coming into the previously unlockable partitions, the request will be unnecessarily delayed. It may not be a big issue, but the consumer fetch doesn't have this problem.

Reply:
Yes, that's right and we are working on a different fetch pattern which should resolve this concern. It will be more inline with exiting fetch and delayed fetches.